subprojects/gradle/module-dependencies-graph/src/gradleTest/kotlin/com/avito/module/dependencies/DependenciesGraphTest.kt

```diff
@@ -1,110 +1,92 @@
-                |:RootB depends on :UniqueB in ConfigurationType.AndroidTests
-                |:RootB depends on :LeafC in ConfigurationType.Main
-                |:RootB Node :LeafG
-                |:LeafG depends on :LeafE in ConfigurationType.Main
-                |:LeafG Node :LeafE
-                |:RootB Node :UniqueB
-                |:RootB Node :LeafC
-                |:LeafC depends on :LeafD in ConfigurationType.Main
-                |:LeafC Node :LeafD
-                |:LeafD depends on :LeafE in ConfigurationType.Main
-                |:LeafD Node :LeafE
-                |print ended""".trimMargin()
-            )
+    fun `avoid pseudo cycles through text fixtures`() {
+        val root = rootProject()
+
+        val lib = androidLib("lib", parent = root)
+        val androidTestLib = androidLib("test-lib", parent = root)
+        androidTestLib.dependencies.add("implementation", lib)
+        lib.dependencies.add("androidTestImplementation", androidTestLib) // depends on lib through implementation
+
+        val app = androidApp("app", parent = root)
+        app.dependencies.add("implementation", lib)
+
+        val modules = graphBuilder(root).buildDependenciesGraph()
+        assertThat(modules).hasSize(3)
+
+        val appNode = modules.find { it.project.path == ":lib" }!!
+
+        val testDependencies = appNode.dependencies(ConfigurationType.AndroidTests)
+        assertThat(testDependencies.size).isEqualTo(1)
+        assertThat(testDependencies.first().project.path).isEqualTo(":test-lib")
     }
 
-    @Test
-    fun `print android_test flatten dependencies graph`() {
-        gradlew(projectDir, "printDependenciesGraph", "--flatten")
-            .assertThat()
-            .buildSuccessful()
-            .outputContains(
-                """
-                |> Task :printDependenciesGraph
-                |Root :CopyRootB
-                |:CopyRootB graph contains :LeafG
-                |:CopyRootB graph contains :LeafE
-                |:CopyRootB graph contains :UniqueCopyB
-                |Root :RootA
-                |:RootA graph contains :LeafE
-                |:RootA graph contains :UniqueA
-                |Root :RootB
-                |:RootB graph contains :LeafG
-                |:RootB graph contains :LeafE
-                |:RootB graph contains :UniqueB
-                |print ended""".trimMargin()
-            )
+    private fun graphBuilder(root: Project): DependenciesGraphBuilder =
+         DependenciesGraphBuilder(root, StubLoggerFactory)
+
+    private fun rootProject(): Project =
+        ProjectBuilder.builder()
+            .withName("root")
+            .build()
+
+    private fun androidApp(name: String, parent: Project): Project {
+        val project = buildProject(name, parent)
+        project.plugins.apply("com.android.application")
+        project.repositories.run {
+            add(mavenCentral())
+            add(google())
+        }
+        return project
     }
```

Comment:
I've used `ProjectBuilder` to test logic. 
It requires to make something kinda TestProjectGenerator. I'll file an issue.

But even in this case, it's worth it. We don't have to wait for slow GradleTestKit and parse the output.